server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java

```diff
@@ -0,0 +1,146 @@
+package org.activityinfo.ui.client.component.form.field.attachment;
+/*
+ * #%L
+ * ActivityInfo Server
+ * %%
+ * Copyright (C) 2009 - 2013 UNICEF
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as
+ * published by the Free Software Foundation, either version 3 of the 
+ * License, or (at your option) any later version.
+ * 
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ * 
+ * You should have received a copy of the GNU General Public 
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/gpl-3.0.html>.
+ * #L%
+ */
+
+import com.google.common.base.Strings;
+import com.google.gwt.http.client.*;
+import com.google.gwt.user.client.Window;
+import com.google.gwt.user.client.ui.FileUpload;
+import com.google.gwt.user.client.ui.FormPanel;
+import com.google.gwt.user.client.ui.Hidden;
+import com.google.gwt.user.client.ui.VerticalPanel;
+import org.activityinfo.core.shared.util.MimeTypeUtil;
+import org.activityinfo.i18n.shared.I18N;
+import org.activityinfo.legacy.shared.Log;
+import org.activityinfo.model.resource.Resource;
+import org.activityinfo.model.resource.ResourceId;
+import org.activityinfo.model.resource.Resources;
+import org.activityinfo.model.type.attachment.Attachment;
+import org.activityinfo.model.util.Holder;
+import org.activityinfo.service.blob.UploadCredentials;
+import org.activityinfo.ui.client.util.GwtUtil;
+
+import javax.annotation.Nullable;
+import java.util.Map;
+
+/**
+ * @author yuriyz on 11/16/2015.
+ */
+public class Uploader {
+
+    public interface UploadCallback {
+        void onFailure(@Nullable Throwable exception);
+
+        void upload();
+    }
+
+    private final FileUpload fileUpload;
+    private final FormPanel formPanel;
+    private final VerticalPanel hiddenFieldsContainer;
+    private final UploadCallback uploadCallback;
+
+    private final Holder<Attachment> attachment;
+
+    public Uploader(FormPanel formPanel, FileUpload fileUpload, Holder<Attachment> attachment,
+                    VerticalPanel hiddenFieldsContainer, UploadCallback uploadCallback) {
+        this.formPanel = formPanel;
+        this.fileUpload = fileUpload;
+        this.attachment = attachment;
+        this.hiddenFieldsContainer = hiddenFieldsContainer;
+        this.uploadCallback = uploadCallback;
+    }
+
+    public void requestUploadUrl() {
+        try {
+            RequestBuilder requestBuilder = new RequestBuilder(RequestBuilder.POST, URL.encode(createUploadUrl()));
```

Comment:
It's url for requesting upload url (which is constructed manually in the same class).